ovmf-vars-generator

```diff
@@ -116,25 +120,46 @@ def enroll_keys(args):
     if args.print_output:
         print(strip_special(read), end='')
         print()
-    # Send the escape char to enter the UEFI shell early
-    p.stdin.write(b'\x1b')
-    p.stdin.flush()
-    # And then run the following three commands from the UEFI shell:
-    # change into the first file system device; install the default
-    # keys and certificates, and reboot
-    p.stdin.write(b'fs0:\r\n')
-    p.stdin.write(b'EnrollDefaultKeys.efi\r\n')
-    p.stdin.write(b'reset -s\r\n')
+    # Send the press enter for the default boot (uefi shell)
+    p.stdin.write(b'\r\n')
     p.stdin.flush()
+
+    wait_timeout = 100;
     while True:
-        read = p.stdout.readline()
+        poll_result = pollobj.poll(1000)
+        if poll_result:
+            # readline can get stuck in menus and other
+            # UI elements in uefi which don't respond
+            # with a CR the poll above doesn't help
+            # with that case.
+            read = p.stdout.readline()
+        else:
+            wait_timeout-=1
+            if wait_timeout == 0:
+                logging.info('Failed enrollment')
+                # consider something stronger here
+                # as we can still get stuck in the p.wait()
+                break;
+            continue
         if args.print_output and len(read):
             print('OUT: %s' % strip_special(read), end='')
             print()
-        if b'info: success' in read:
+        if b'seconds to skip' in read:
+            p.stdin.write(b'\r\n')
+            p.stdin.flush()
+        elif b'info: success' in read:
             break
         elif b'Reset with <null string>' in read:
             break
+        elif b'Shell>' in read:
+            # And then run the following three commands from the UEFI shell:
+            # change into the first file system device; install the default
+            # keys and certificates, and reboot
+            p.stdin.write(b'fs0:\r\n')
+            p.stdin.write(b'EnrollDefaultKeys.efi\r\n')
+            p.stdin.write(b'reset -s\r\n')
+            p.stdin.flush()
+
```

Comment:
fwiw, I've had good luck with pexpect: https://salsa.debian.org/qemu-team/edk2/-/blob/debian/debian/tests/shell.py

Reply:
Which serves a purpose if one is in a python env. OTOH, python isn't a great shell scripting language anymore than c. Worse, this script is being used on fedora in an actual rpmbuild/bash environment.  In that case about 80% of this scripts functionality can be represented like:

```
/usr/bin/expect <<EOF
spawn /usr/bin/qemu-system-${MACHTYPE} ${QEMU_ARCH} ${COMMON} $FIRMWARE ${VARS}=off $CDROM $SMBIOS
expect -exact "Shell>"
send -- "\r\nfs0:\r\n"
send -- "EnrollDefaultKeys.efi\r\n"
send -- "reset -s\r\n"
expect eof
EOF

```
the remainder, which I've also done (dealing with multiple arches, and validating enrolled keys) is about 3-4x the size of that.

Reply:
> Which serves a purpose if one is in a python env. OTOH, python isn't a great shell scripting language anymore than c. Worse, this script is being used on fedora in an actual rpmbuild/bash environment.

I'm not arguing for/against this being in `python` vs. `sh`, just that since it *is* in `python`, `pexpect` is at our disposal.